android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/BaseQuestionnaireFragment.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2021 Ona Systems, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.ui.questionnaire
+
+import com.google.android.fhir.datacapture.QuestionnaireFragment
+import com.google.android.fhir.datacapture.contrib.QuestionnaireItemBarCodeReaderViewHolderFactory
+
+class BaseQuestionnaireFragment : QuestionnaireFragment() {
+
+  override fun getCustomQuestionnaireItemViewHolderFactoryMatchers():
```

Comment:
not sure I understand this, can you explain what this is doing?

Reply:
In SDC we don't have any specific type related to the barcode widget. Therefore, if we need to register any custom type in the SDC widget factory so in that case, we can register it by using the custom extension. we can add the extension in the questionnaire JSON and register in `getCustomQuestionnaireItemViewHolderFactoryMatchers()` method by filtering the extension URL.

here is an example questionnaire JSON snippet.

Reply:
OK I see, thank you for the explanation. Two questions:

1) are there situations where the `url` will be the barcode widget, but the `valueString` will not be `"barcode"`?
2) it looks like this function is only returning barcode widgets, should this function name be specific to that, something like the suggested? Alternatively, should the url and the value string be parameters passed to this function? With the name remaining as it is
```suggestion
  override fun getBarcodeQuestionnaireItemViewHolderFactoryMatchers():
```

Reply:
for point 1. both `URL` and `valueString` will be required to resolve the custom widget and the value could be anything but both values i.e `URL` and `valueString` should match the questionnaire item values. you can check the below code snippet.

https://github.com/opensrp/fhircore/blob/93faf45b9e6e31311050f8a92c0ecd0c21007fa6/android/engine/src/main/java/org/smartregister/fhircore/engine/ui/questionnaire/BaseQuestionnaireFragment.kt#L29-L32

For point 2. We can't change the method name nor pass the argument because it's an overridden method and coming from the SDC library.

Reply:
ah right an override, so can't change signature, what I'm confused about is that the name makes this sound general, but the way we are overriding it, it's now specific to only barcode widgets.

looking forward, suppose we have a map widget we can add to a questionnaire, how would we change this class (or another) to load that widget?

Reply:
yes we can add more widgets, as we are returning the list of `QuestionnaireItemViewHolderFactoryMatcher`. checkout this file https://github.com/google/android-fhir/blob/4c939d3796c231e65d7ef477539e92d46ecbb782/datacapturegallery/src/main/java/com/google/android/fhir/datacapture/gallery/CustomQuestionnaireFragment.kt here we are adding two custom widgets `number-picker` and `barcode`

Reply:
As @owais-vd mentioned the [QuestionnaireFragment#getCustomQuestionnaireItemViewHolderFactoryMatchers](https://github.com/google/android-fhir/blob/d3c3feda998e7c7fdc20c07b539247c892c0fb72/datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt#L94) function signature is from the SDK and can only be modified from there. From the code documentation, the intention was to generalize the method such that the implementer can provide a list of custom views required for their use case. So @owais-vd implementation is correct. The only alternative is to include the widgets as part of the SDK. I do not think extending the `QuestionnaireFragment` from the `engine` module is necessary, PS we conventionally name abstract classes with the prefix `Base` so as to not confuse extended classes.

Reply:
ah cool, got it, so we add them to the list then filter again later on